### PR TITLE
Add overlays to the debugger.

### PR DESCRIPTION
**Pull request by Malabarba**
Too much? :)  
### end-of-line
cider-debug-overlay-2

### inline
cider-debug-overlays

### CODE REVIEW

**Comment by Malabarba** [comment]
I figured this is safe enough to go in `0.9.1`, so it's based of `master`.

**Comment by bbatsov** [comment]
I like it, although it should probably be configurable, as people used to `edebug` might prefer the use of the minibuffer. Btw, it'd be great if you expand the use of overlays for interactive evaluation as well. This will make it possible for us to do something like `instarepl` / `litable` long term.

**Comment by Malabarba** [comment]
>  I like it, although it should probably be configurable, as people used to edebug might prefer the use of the minibuffer. 

Oh, the screenshot doesn't show, but it still uses the minibuffer. And there's a custom option to turn the overlays off.

> Btw, it'd be great if you expand the use of overlays for interactive evaluation as well. This will make it possible for us to do something like instarepl / litable long term.

The only problem is that these overlays are a little annoying if they stick around. In order to not change the contents of the buffer, they have 0 width, which means the user is completely unable to erase them. There's probably a better way to do this though, as I'm sure I've seen somewhere overlays that vanish when you write over them.

**Comment by bbatsov** [comment]
> There's probably a better way to do this though, as I'm sure I've seen somewhere overlays that vanish when you write over them.

Yep, that'd be best. Another thing to consider is their exact position - some people might not like the overlay shifting things to the right, so an alternative showing the overlay outside the text on the line might be nice to have as well.

File: cider-debug.el [hunk 1]
```diff
@@ -82,9 +82,60 @@ sexp."
...
+                 'face 'cider-debug-code-overlay-face)
+    ;; Higher priority than `show-paren'.
+    (overlay-put cider--debug-code-overlay 'priority 2000)))
+
+(defcustom cider-debug-use-overlays t
+  "Whether to higlight debugging information with overlays.
+Only applies to \"*cider-debug ...*\" buffers, which are used in debugging
+sessions."
+  :type 'boolean
+  :group 'cider)
```
Comment (bbatsov) [inline cider-debug.el]: Don't forget the `:package-version` properties.

**Comment by bbatsov** [comment]
On an unrelated note - add a changelog entry for this and drop the `.` from the commit message's title.

**Comment by Malabarba** [comment]
All done, and I've factored out the overlay handling as well so it can be reused. Should these functions go in another file or do I leave them there?

**Comment by Malabarba** [comment]
Also, see new pic at the top.

**Comment by bbatsov** [comment]
:+1: Fantastic work!

**Comment by arrdem** [comment]
:+1:

**Comment by mavbozo** [comment]
:thumbsup: Awesome!!

**Pull request merged**